toolkit.nu

```diff
@@ -31,36 +33,37 @@ export def fmt [
 # > it is important to make `clippy` happy :relieved:
 export def clippy [
     --verbose: bool # print extra information about the command's progress
-    --dataframe: bool # use the dataframe feature
+    --features: list<string> # the list of features to run *Clippy* on
 ] {
     if $verbose {
         print $"running ('toolkit clippy' | pretty-print-command)"
     }
 
-    try {
-        if $dataframe {
-            cargo clippy --workspace --features=dataframe,extra -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect -A clippy::result_large_err
-        } else {
-            cargo clippy --workspace -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect -A clippy::result_large_err
+    try {(
+        cargo clippy
+            --workspace
+            --features ($features | str join ",")
```

Comment:
let me try :wink:

Reply:
i apply the following patch to have clear *Clippy* errors in both the `dataframe` and the `extra` feature :+1: 
```diff
diff --git a/crates/nu-cmd-dataframe/src/dataframe/lazy/select.rs b/crates/nu-cmd-dataframe/src/dataframe/lazy/select.rs
index 2d67cfc2e..2ac0ae706 100644
--- a/crates/nu-cmd-dataframe/src/dataframe/lazy/select.rs
+++ b/crates/nu-cmd-dataframe/src/dataframe/lazy/select.rs
@@ -55,7 +55,7 @@ impl Command for LazySelect {
         let vals: Vec<Value> = call.rest(engine_state, stack, 0)?;
         let value = Value::List {
             vals,
-            span: call.head,
+            span: call.head.clone(),
         };
         let expressions = NuExpression::extract_exprs(value)?;
 
diff --git a/crates/nu-cmd-extra/src/extra/bits/or.rs b/crates/nu-cmd-extra/src/extra/bits/or.rs
index 1c3c7243d..2846be4c7 100644
--- a/crates/nu-cmd-extra/src/extra/bits/or.rs
+++ b/crates/nu-cmd-extra/src/extra/bits/or.rs
@@ -40,7 +40,7 @@ impl Command for BitsOr {
         call: &Call,
         input: PipelineData,
     ) -> Result<PipelineData, ShellError> {
-        let head = call.head;
+        let head = call.head.clone();
         let target: i64 = call.req(engine_state, stack, 0)?;
 
         // This doesn't match explicit nulls
```
> **Note**  
> see the useless `.clone()` :smirk: 

in the following i'll run the following command, `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect`, just adding `--features ...` to `cargo clippy` in a bunch of cases :ok_hand: 

- no `--features` option: :x: extra :x: dataframe
- `--features ""`: :x: extra :x: dataframe
- `--features dataframe`: :x: extra :x: dataframe
- `--features extra`: :x: extra :x: dataframe
- `--features dataframe,extra`: :x: extra :x: dataframe

:postal_horn: i'm starting to wonder if this option makes any difference :sweat_smile:

Reply:
however, if i remove the options after `--`, i only get :warning: in all the cases above :thinking:

Reply:
i think `--workspace` bypasses the `--features` option :open_mouth: 

because the features are members of the workspace, triggering `--workspace` is much stronger than `--features` :smirk: 

if i conduct exactly the same experiments as above, just by removing the `--workspace` option, i get the following:
- no `--features` option: :heavy_check_mark: extra :heavy_check_mark: dataframe
- `--features ""`: :heavy_check_mark: extra :heavy_check_mark: dataframe
- `--features dataframe`: :heavy_check_mark: extra :x: dataframe
- `--features extra`: :x: extra :heavy_check_mark: dataframe
- `--features dataframe,extra`: :x: extra :x: dataframe

Reply:
what do you think we should do here?

we need to think about the CI and the PR template too with this whole cratification :thinking:

Reply:
It worked the way it was before so you could have something like
if $features | is-empty
  don't use --features
else
  use --features

I actually never use the --workspace argument at all

Reply:
the issue is not with `--features` here.
as shown above, without `--workspace`, `--features ($features | str join ",")` will not compile any additional feature when the `$features` list is empty!

> you can apply the patch above which introduces *Clippy* errors and try the following function
> ```nu
> def foo [features?: list<string>] {(
>     cargo clippy
>         --features ($features | default [] | str join ",")
>     --
>         -D warnings
>         -D clippy::unwrap_used
>         -A clippy::needless_collect
> )}
> ```

Reply:
because i do not think it was working before :thinking:

Reply:
yes, i don't understand or believe we need the --workspace flag. Like I said, I never use it.

Reply:
oook i got you covered :smirk: 

commit incoming :muscle:

Reply:
should be a lot more meaningful in d0452157e :relieved:

Reply:
what's the point of --workspace?

Reply:
to run the commands on the whole workspace, i.e. all the features, all the items, it overrides things like `--features` :+1:

Reply:
we might use `--workspace` in the CI for instance :open_mouth:

Reply:
yeah, [here](https://github.com/nushell/nushell/blob/main/.github/workflows/ci.yml#L49) :yum:

Reply:
That's great but it doesn't explain to me why it's needed. I'll have to google it.

Reply:
because i understand that by default, i.e. without any option, things like `cargo build` will only apply on the *default workspace members*, so i'd say not *all* of them like explicitely with the `--workspace` option.

Reply:
@fdncred 
just a friendly reminder of this PR :relieved: 
**have you had the time to look at the `--workspace` option?**

> **Note**
> i've completed the experiment [here](https://github.com/nushell/nushell/issues/9439#issuecomment-1594367915) :+1:

Reply:
Like I said before, I never use the --workspace flag at all. For my personal use, I feel that it is useless. Nothing said so far in this thread has changed my mind. Having said that, I don't care what the options are to compile things. I just want it to work. Right now, this is the command I always use on Windows, MacOS, and WSL2/Ubuntu `cargo install --path . --features=dataframe,extra,mimalloc` and it works the way I need it to.

Reply:
then you should be able to do `toolkit install --features dataframe extra mimalloc` :+1: 

it's just that the toolkit supports both `--workspace` and `--features`: afaik `--workspace` is just a shortcut to "all the features" so that you do not have to specify them all :yum:

Reply:
is that ok with you? :innocent:

Reply:
as long as it works, i'm ok with it. i'd also want to make sure the `make all` functionality that builds all the plugins, doesn't use those features, but the building of nushell does. but feel free to land it, if that's the case.

Reply:
this PR does not change these plugin commands at all, so i think we can land it :relieved: 

and in the end, i think it even works as you expect right now
- Nushell is built with whatever feature the user passes [here](https://github.com/nushell/nushell/blob/main/toolkit.nu#L326)
- plugins are not built with any feature specified [here](https://github.com/nushell/nushell/blob/main/toolkit.nu#L312-L319)